mods/gen2/moves.js

```diff
@@ -12,13 +12,17 @@ exports.BattleMovedex = {
 			if (target.boosts.atk >= 6) {
 				return false;
 			}
+
+			var boost = 0;
+
 			if (target.hp <= target.maxhp / 2) {
-				this.boost({atk: 2});
-				return false;
+				boost = (target.boost ? target.boost.atk : 0) + 2;
```

Comment:
There is a move routine that is called by both Swords Dance and Belly Drum. This routine increments the attack stage by 2 and returns. What the actual game code does is a ``do-while``, calls that routine while the attack stage is below six, breaking if it finds that the HP of the user is below floor(maxhp  / 2).

For more information, there's the RoA research threads for gens 1 and 2, where the relevant part of the disassembly and how it works is often quoted and explained (most notoriously by @xCrystal).